### PR TITLE
Feature/aligned ends 2

### PR DESCRIPTION
**Pull request by rrahn**
### Supersedes:
* ~#660~ 

### Dependencies:
 * ~#657~

### Info
 * Adapts the configurator to reslolve the aligned_ends configuration.
 * Adds a traits class to the affine_gap_init_policy to select the proper initiailisation.
 * Implements a find_optimum_policy which tracks the optimal score depending on the aligned_ends configuration.
 * Adapts the alignment algorithm to check for the optimal score.
 * Fixes the CRTP-Base class pattern using private member and befirending the derived class.
 * Fixes the CRTP-Base class declaration.
 * Updates documentation of the classes.

### CODE REVIEW

File: include/seqan3/alignment/pairwise/alignment_configurator.hpp [hunk 1]
```diff
@@ -157,33 +165,196 @@ struct alignment_configurator
...
+
+    // ----------------------------------------------------------------------------
+    // configure initialization policy
+    // ----------------------------------------------------------------------------
+
+    // Get the value for the sequence ends configuration.
+    auto align_ends_cfg = cfg.template value_or<align_cfg::aligned_ends>(align_cfg::none_ends_free);
+    using align_ends_cfg_t = decltype(align_ends_cfg);
+
+    // This lambda augments the initialization policy of the alignment algorithm
```
Comment (eseiler) [inline include/seqan3/alignment/pairwise/alignment_configurator.hpp]:
```suggestion
    // This lambda augments the initialisation policy of the alignment algorithm
```

Also everywhere else if accepted

File: include/seqan3/alignment/pairwise/policy/affine_gap_init_policy.hpp [hunk 2]
```diff
@@ -14,20 +14,40 @@
...
 
 namespace seqan3::detail
 {
 
+/*!\brief The default traits class for seqan3::detail::affine_gap_init_policy.
+ * \ingroup alignment_policy
+ *
+ * \details
+ *
+ * Enables the behavior of a global alignment where both sides of the dynamic programming matrix are
```
Comment (eseiler) [inline include/seqan3/alignment/pairwise/policy/affine_gap_init_policy.hpp]:
```suggestion
 * Enables the behaviour of a global alignment where both sides of the dynamic programming matrix are
```

File: include/seqan3/alignment/pairwise/policy/find_optimum_policy.hpp [hunk 3]
```diff
@@ -0,0 +1,123 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2018, Knut Reinert & Freie Universität Berlin
```
Comment (eseiler) [inline include/seqan3/alignment/pairwise/policy/find_optimum_policy.hpp]:
```suggestion
// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
```

File: include/seqan3/alignment/pairwise/policy/find_optimum_policy.hpp [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+// Copyright (c) 2016-2018, Knut Reinert & MPI für molekulare Genetik
```
Comment (eseiler) [inline include/seqan3/alignment/pairwise/policy/find_optimum_policy.hpp]:
```suggestion
// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
```

File: include/seqan3/alignment/pairwise/policy/find_optimum_policy.hpp [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+namespace seqan3::detail
+{
+
+/*!\brief The default traits class for seqan3::detail::find_optimum_policy.
+ * \ingroup alignment_policy
+ *
+ * \details
+ *
+ * Defines the behavior of a global alignment, in which only the last cell of the dynamic programming matrix is
```
Comment (eseiler) [inline include/seqan3/alignment/pairwise/policy/find_optimum_policy.hpp]:
```suggestion
 * Defines the behaviour of a global alignment, in which only the last cell of the dynamic programming matrix is
```

File: include/seqan3/alignment/pairwise/policy/find_optimum_policy.hpp [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    template <typename score_t>
+    constexpr void check_score_last_row([[maybe_unused]] score_t const val,
+                                        [[maybe_unused]] score_t & optimum) const noexcept
+    {
+        if constexpr (traits_type::find_in_last_row_t::value)
+            optimum = std::max(optimum, val);
+    }
+
+    /*!\brief Checks the complete last column for the optimal score.
+     * \tparam rng_t    The type of the last column. Must be at least an input range.
```
Comment (eseiler) [inline include/seqan3/alignment/pairwise/policy/find_optimum_policy.hpp]: Mhm if it needs to be an input range, you could also say `models std::InputRange` and require it?

File: include/seqan3/alignment/pairwise/policy/find_optimum_policy.hpp [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        using std::get;
+        // Only check the entire column if it was configured to search here.
+        if constexpr (traits_type::find_in_last_column_t::value)
+        {
+            ranges::for_each(rng, [&](auto && tpl)
+            {
+                optimum = std::max(optimum, get<0>(tpl));
+            });
+        }
+        else  // Only check the last cell of the global alignment.
```
Comment (eseiler) [inline include/seqan3/alignment/pairwise/policy/find_optimum_policy.hpp]: So this is the default when no policies are set?

File: include/seqan3/alignment/pairwise/policy/find_optimum_policy.hpp [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+     */
+    template <typename score_t>
+    constexpr void check_score([[maybe_unused]] score_t const val,
+                               [[maybe_unused]] score_t & optimum) const noexcept
+    {
+        if constexpr (traits_type::find_in_every_cell_t::value)
+            optimum = std::max(optimum, val);
+    }
+
+    /*!\brief Checks the last cell of the last row of the dynamic programming matrix.
```
Comment (eseiler) [inline include/seqan3/alignment/pairwise/policy/find_optimum_policy.hpp]: The name and trait suggest that every cell in the last row is checked.

File: include/seqan3/alignment/pairwise/alignment_configurator.hpp [hunk 9]
```diff
@@ -157,33 +165,196 @@ struct alignment_configurator
...
+    static constexpr auto configure_free_ends(config_t const &);
+
+    /*!\brief Configures the find_optimum_policy for the alignment algorithm.
+     * \tparam kernel_t     The type-erased kernel to return.
+     * \tparam gap_policy_t The type of the core alignment policy needed for the find_optimum_policy.
+     * \tparam policies_t   A template parameter pack for the remaining policies.
+     * \tparam config_t     The alignment configuration type.
+     * \param[in] cfg       The passed configuration object.
+     *
+     * \details
```
[the remark this answers is not in the document]
Comment (rrahn) [reply include/seqan3/alignment/pairwise/alignment_configurator.hpp]: We agreed on making it explicit in the style guides

File: include/seqan3/alignment/pairwise/alignment_configurator.hpp [hunk 10]
```diff
@@ -0,0 +1,360 @@
...
+     * Thus the core alignment policy computing the cells must be able to access the functions in the
+     * find_optimum_policy. Since both classes are crtp base classes of the alignment algorithm the find_optimum_policy
+     * must befriend the core_policy_t in order to grant it access to the contained member functions.
+     */
+    template <typename kernel_t, typename core_policy_t, typename ...policies_t, typename config_t>
+    static constexpr auto configure_find_optimum(config_t const &);
+
+};
+
+//!\cond
```
[the remark this answers is not in the document]
Comment (rrahn) [reply include/seqan3/alignment/pairwise/alignment_configurator.hpp]: No, this is only the defintion. The declaration of the function is documented.

File: include/seqan3/alignment/pairwise/policy/find_optimum_policy.hpp (hunk 6 above)
Comment (rrahn) [reply include/seqan3/alignment/pairwise/policy/find_optimum_policy.hpp]: It is actually required to be a bidirectional range and I added a concept.

File: include/seqan3/alignment/pairwise/policy/find_optimum_policy.hpp (hunk 8 above)
Comment (rrahn) [reply include/seqan3/alignment/pairwise/policy/find_optimum_policy.hpp]: I clarified the behaviour in the documentation.

**Pull request closed without merging**